src/sentry/models/featureadoption.py

```diff
@@ -0,0 +1,173 @@
+# TODO(ehfeng) manager.add("merge", "Merge Issues", "web", prerequisite=["first_event"])
+# TODO(ehfeng) manager.add("releases", "Releases", "web", prerequisite=["first_project"])
+
+# Admin UI
+manager.add(80, "sso", "SSO", "admin", prerequisite=["invite_team"])
+manager.add(81, "data_scrubbers", "Data Scrubbers", "admin", prerequisite=["first_event"])
+
+
+class FeatureAdoptionManager(BaseManager):
+
+    @staticmethod
+    def in_cache(organization_id, feature_id):
+        org_key = FEATURE_ADOPTION_REDIS_KEY.format(organization_id)
+        feature_matches = []
+        with redis.clusters.get('default').map() as client:
+            feature_matches.append(client.sismember(org_key, feature_id))
+
+        return any([p.value for p in feature_matches])
+
+    @staticmethod
+    def set_cache(organization_id, feature_id):
+        org_key = FEATURE_ADOPTION_REDIS_KEY.format(organization_id)
+        with redis.clusters.get('default').map() as client:
+            client.sadd(org_key, feature_id)
+
+    @staticmethod
+    def get_all_cache(organization_id):
+        org_key = FEATURE_ADOPTION_REDIS_KEY.format(organization_id)
+        result = []
+        with redis.clusters.get('default').map() as client:
+            result.append(client.smembers(org_key))
+
+        return {int(x) for x in set.union(*[p.value for p in result])}
+
+    @staticmethod
+    def bulk_set_cache(organization_id, *args):
+        org_key = FEATURE_ADOPTION_REDIS_KEY.format(organization_id)
+        with redis.clusters.get('default').map() as client:
+            client.sadd(org_key, *args)
+
+    def record(self, organization_id, feature_slug, **kwargs):
+        try:
+            feature_id = manager.get_by_slug(feature_slug).id
+        except KeyError:
+            logger.info('Invalid feature slug: %s' % feature_slug)
+            return
+
+        if not self.in_cache(organization_id, feature_id):
+            row, created = self.create_or_update(
+                organization_id=organization_id,
+                feature_id=feature_id,
+                complete=True)
+            self.set_cache(organization_id, feature_id)
+            return created
+
+        return False
+
+    def bulk_record(self, organization_id, feature_slugs, **kwargs):
+        try:
+            features = []
+            incomplete_feature_ids = set([manager.get_by_slug(slug).id for slug in feature_slugs]) - self.get_all_cache(organization_id)
+
+            for feature_id in incomplete_feature_ids:
+                features.append(FeatureAdoption(
+                    organization_id=organization_id,
+                    feature_id=feature_id,
+                    complete=True))
+
+            self.bulk_create(features)
+            self.bulk_set_cache(organization_id, *incomplete_feature_ids)
+
+        except KeyError:
+            logger.info('Invalid feature slug: %s' % feature_slugs)
+            return
+        except IntegrityError:
```

Comment:
`self.bulk_create(features)` would throw this.

I'm torn on whether to include the `self.bulk_set_cache` line in a `finally` or just after the `try` block. Effectively, in the case where features 1 and 2 are done (but Redis has failed), and we try and bulk insert features 1, 2, and 3, if the IntegrityError occurs, do we still want to update the Redis cache? Otherwise, the IntegrityError will keep happening and logging. But if we do update Redis (in the `finally` block), it'll log the IntegrityError once but be silent after that.